### ISSUE TITLE
Rails 5 : Object locking doesn't work: Combination of limit and lock is not supported.

### ISSUE THREAD

**Issue by Envek** [rsim/oracle-enhanced#920]
Code like:

``` rb
address = Address.first
address.lock!
```

Doesn't work anymore on _any model_ in our app after upgrade to Rails 5.

```
Exception: ArgumentError:             'Combination of limit and lock is not supported.
            because generated SQL statements
            `SELECT FOR UPDATE and FETCH FIRST n ROWS` generates ORA-02014.`

--
 0: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/arel-7.1.0/lib/arel/visitors/oracle12.rb:9:in `visit_Arel_Nodes_SelectStatement'
 1: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/arel-7.1.0/lib/arel/visitors/reduce.rb:13:in `visit'
 2: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/arel-7.1.0/lib/arel/visitors/reduce.rb:7:in `accept'
 3: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/determine_if_preparable_visitor.rb:8:in `accept'
 4: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/database_statements.rb:12:in `to_sql'
 5: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/database_statements.rb:32:in `select_all'
 6: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/connection_adapters/abstract/query_cache.rb:70:in `select_all'
 7: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/querying.rb:39:in `find_by_sql'
 8: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation.rb:699:in `exec_queries'
 9: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation.rb:580:in `load'
10: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation.rb:260:in `records'
11: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation/finder_methods.rb:525:in `find_take'
12: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation/finder_methods.rb:100:in `take'
13: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation/finder_methods.rb:475:in `find_one'
14: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation/finder_methods.rb:456:in `find_with_ids'
15: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation/finder_methods.rb:66:in `find'
16: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/persistence.rb:437:in `block in reload'
17: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/scoping/default.rb:35:in `block in unscoped'
18: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/relation.rb:351:in `scoping'
19: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/scoping/default.rb:35:in `unscoped'
20: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/persistence.rb:437:in `reload'
21: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/attribute_methods/dirty.rb:37:in `reload'
22: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/associations.rb:253:in `reload'
23: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/autosave_association.rb:235:in `reload'
24: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/aggregations.rb:13:in `reload'
25: /Users/envek/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/activerecord-5.0.0/lib/active_record/locking/pessimistic.rb:62:in `lock!'
26: (pry):5:in `<main>'
```

Ruby MRI 2.3.1
Ruby on Rails 5.0.0
Oracle Enhanced Adapter 1.7.0.beta1
Arel 7.1.0
Oracle 12c from https://github.com/MaksymBilenko/docker-oracle-12c

Thank you for continuing development of this awesome adapter!

**Comment by yahonda** [comment]
Thanks for testing 1.7.0.beta1. This is an expected behavior based on https://github.com/rails/arel/pull/430

Here are background for this pull request:
https://github.com/rails/rails/issues/24779
https://github.com/rails/arel/pull/337#discussion_r61656330

**Comment by Envek** [comment]
Is it possible to change query not to use limit as we have only one record with it's primary key so we can just do `SELECT FOR UPDATE "MODELS".* FROM "MODELS" WHERE "MODELS"."ID" = :a1`?

**Comment by yahonda** [comment]
Unfortunately I do not think we can change it from framework code, Arel and/or Oracle enhanced adapter.  

If I understand it correctly `first` is associated with `limit` whichever it is used with `lock`.
Application code needs change to eliminate rows by specifying its id (or any other condition) not `first`.

**Comment by yahonda** [comment]
Of rouce there may be some alternative then it would be helpful for you to refer https://github.com/rails/arel/blob/master/lib/arel/visitors/oracle12.rb and open pull requests to Arel if you have any idea.

**Comment by rlafranchi** [comment]
Is there any progress on this?  I came across this issue in attempts to upgrade to rails 5.  It also doesn't work when just running `find(id)` which doesn't make sense.  It seems as though the "FETCH FIRST" portion of the SQL should be removed when find is used.

**Comment by yahonda** [comment]
#1049 has been merged to master which allows Rails 5 and Oracle 12 database users to fall back to old visitor. Then users can workaround this behavior.

```ruby
ActiveRecord::ConnectionAdapters::OracleEnhancedAdapter.use_old_oracle_visitor = true
```
I understand this is not an identical solution, but this is the best thing that I can do now.

**Comment by yahonda** [comment]
Closing since Oracle enhanced adapter 1.7.7 has been released including #1049.

**Comment by remofritzsche** [comment]
Even with `use_old_oracle_visitor` set to `true` it doesn't work:

```
pry(main)> ActiveRecord::ConnectionAdapters::OracleEnhancedAdapter.use_old_oracle_visitor
=> true
pry(main)> u = User.first.lock!
  User Load (106.0ms)  SELECT  "USERS".* FROM "USERS" ORDER BY "USERS"."ID" ASC FETCH FIRST :a1 ROWS ONLY  [["LIMIT", 1]]
ArgumentError:             'Combination of limit and lock is not supported.
            because generated SQL statements
            `SELECT FOR UPDATE and FETCH FIRST n ROWS` generates ORA-02014.`
from /work/naz/eld/vendor/gems/gems/arel-7.1.4/lib/arel/visitors/oracle12.rb:9:in `visit_Arel_Nodes_SelectStatement'
```

The way I see it, locking simply does't work anymore with Oracle and Rails 5 in most cases. I'd be thrilled if there was a solution for this. Thanks a bunch for having a look at this :)
